polyfill/lib/calendar.mjs

```diff
@@ -230,15 +230,55 @@ export class Calendar {
     date = ES.ToTemporalDate(date);
     return impl[GetSlot(this, CALENDAR_ID)].dayOfYear(date);
   }
-  weekOfYear(date) {
+  calendarDateWeekOfYear(date) {
+    // Supports only Gregorian and ISO8601 calendar; can be updated to add support for other calendars.
+    // Returns undefined for calendars without a well-defined week calendar system.
+    // eslint-disable-next-line max-len
+    // Also see: https://github.com/unicode-org/icu/blob/ab72ab1d4a3c3f9beeb7d92b0c7817ca93dfdb04/icu4c/source/i18n/calendar.cpp#L1606
     if (!ES.IsTemporalCalendar(this)) throw new TypeError('invalid receiver');
     date = ES.ToTemporalDate(date);
-    return impl[GetSlot(this, CALENDAR_ID)].weekOfYear(date);
+    const id = GetSlot(this, CALENDAR_ID);
+    if (id === 'gregory' || id === 'iso8601') {
+      const calendar = impl[GetSlot(this, CALENDAR_ID)];
+      let yow = GetSlot(date, ISO_YEAR);
+      const dayOfWeek = this.dayOfWeek(date);
+      const dayOfYear = this.dayOfYear(date);
+      const fdow = id === 'iso8601' ? 1 : calendar.helper.getFirstDayOfWeek();
+      const mdow = id === 'iso8601' ? 4 : calendar.helper.getMinimalDaysInFirstWeek();
+      var relDow = (dayOfWeek + 7 - fdow) % 7;
+      // Assuming the year length is less than 7000 days.
+      var relDowJan1 = (dayOfWeek - dayOfYear + 7001 - fdow) % 7;
```

Comment:
This is rather opaque; I can see that in iso8601 `relDowJan1` is 0 in iso8601 for a Monday January 1 and any later day of that year, 1 for a Tuesday January 1 and the rest of that year, etc. but reading meaning from that seems unnecessarily difficult.
```suggestion
      // For both the input date and the first day of its calendar year, calculate the day of week
      // relative to first day of week in the relevant calendar (e.g., in iso8601, relative to Monday).
      var relDow = (dayOfWeek + 7 - fdow) % 7;
      var relDowJan1 = (dayOfWeek - dayOfYear + 7001 - fdow) % 7;
```